### PR TITLE
[Core] Raise an exception on Ray version mismatch

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Mismatching Ray versions will most likely cause issues because we don't have stable API behaviors across internal APIs (inbetween Ray components).

That says, we should better raising an exception rather than making it work. 

TODO
- [x] Unit tests

## Related issue number

https://github.com/ray-project/ray/issues/20166

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_failure.py [hunk 1]
```diff
@@ -448,21 +448,20 @@ def put_task():
...
     ray_version = ray.__version__
-    ray.__version__ = "fake ray version"
-
-    ray.init(num_cpus=1)
-    p = error_pubsub
+    try:
+        cluster = ray_start_cluster
+        cluster.add_node(num_cpus=1)
 
-    errors = get_error_message(p, 1, ray_constants.VERSION_MISMATCH_PUSH_ERROR)
```
Comment (rkooo567) [inline python/ray/tests/test_failure.py:459]: This was not tested at all before (look assert False)

**Comment by rkooo567** [comment]
We also have a feature that if this happens on a worker process, we prints the error. I tried testing it, but there was no easy way to achieve this unfortunately... :(

File: python/ray/_private/services.py [hunk 2]
```diff
@@ -803,10 +803,7 @@ def check_version_info(redis_client):
                          " was started with:" + "\n"
                          "    Ray: " + version_info[0] + "\n"
                          "    Python: " + version_info[1] + "\n")
-        if version_info[:2] != true_version_info[:2]:
-            raise RuntimeError(error_message)
-        else:
-            logger.warning(error_message)
+        raise RuntimeError(error_message)
```
Comment (scv119) [inline python/ray/_private/services.py:806]: cc @ericl @raulchen  do we support any kind of rolling update?

**Review by scv119 — APPROVED** [review]
looks good to me. but better wait for ant and other's confirm.

**Comment by rkooo567** [comment]
This will be merged after the initial branch cut.

**Comment by rkooo567** [comment]
@scv119 should we also compare the commit? (not just the version)?

**Comment by rkooo567** [comment]
That seems to be safer

**Comment by rkooo567** [comment]
Please refer to https://github.com/ray-project/ray/issues/20166#issuecomment-976583427

**Comment by rkooo567** [comment]
Will merge this (only adding tests) and then I will keep tracking of redis removal process to make this work eventually.

**Pull request merged**